core/trino-main/src/main/java/io/trino/connector/system/MaterializedViewSystemTable.java

```diff
@@ -0,0 +1,139 @@
+import static io.trino.metadata.MetadataListing.getMaterializedViews;
+import static io.trino.metadata.MetadataListing.listCatalogs;
+import static io.trino.metadata.MetadataListing.listMaterializedViews;
+import static io.trino.metadata.MetadataListing.listSchemas;
+import static io.trino.spi.connector.SystemTable.Distribution.SINGLE_COORDINATOR;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewSystemTable
+        implements SystemTable
+{
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+    private final MaterializedViewRowAdapter materializedViewAdapter;
+
+    @Inject
+    public MaterializedViewSystemTable(
+            Metadata metadata,
+            AccessControl accessControl,
+            MaterializedViewRowAdapter materializedViewAdapter)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.materializedViewAdapter = requireNonNull(materializedViewAdapter, "materializedViewAdapter is null");
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return materializedViewAdapter.getTableLayout();
+    }
+
+    @Override
+    public RecordCursor cursor(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession connectorSession,
+            TupleDomain<Integer> constraint)
+    {
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+
+        List<QualifiedObjectName> materializedViewNames = getAllMaterializedViewNames(session);
+        List<MaterializedViewDto> materializedViews = getMaterializedViewsByName(session, materializedViewNames);
+        List<Object[]> materializedViewRows = materializedViewAdapter.toTableRows(materializedViews);
+
+        InMemoryRecordSet.Builder displayTable = InMemoryRecordSet.builder(getTableMetadata());
+        materializedViewRows.forEach(displayTable::addRow);
+
+        return displayTable.build().cursor();
+    }
+
+    private List<QualifiedObjectName> getAllMaterializedViewNames(Session session)
+    {
+        ImmutableList.Builder<QualifiedObjectName> materializedViewNames = ImmutableList.builder();
+        Set<String> catalogNames = listCatalogs(session, metadata, accessControl).keySet();
+
+        for (String catalogName : catalogNames) {
+            Set<String> schemaNamesInCatalog = listSchemas(session, metadata, accessControl, catalogName);
+
+            for (String schemaName : schemaNamesInCatalog) {
+                QualifiedTablePrefix materializedViewPrefix = new QualifiedTablePrefix(catalogName, schemaName);
+                materializedViewNames.addAll(getMaterializedViewNamesInSchema(session, materializedViewPrefix));
```

Comment:
We should not need 2 steps of listing names and then the MV definitions.
We should directly use getMaterializedViews on the required prefix.